neural_compressor/torch/algorithms/weight_only/save_load.py

```diff
@@ -37,14 +38,516 @@ def save(model, output_dir="./saved_results"):
+                    _is_autoround = True
+                module_quantization_config = [config for config in q_config_value.values()][0]
+
+        if isinstance(module, torch.nn.Linear):
+            # module without qweight means it is not quantized, then skip it
+            loaded_state_dict_keys_set = set(loaded_state_dict_keys)
+            if (
+                name + ".qweight" not in loaded_state_dict_keys_set
+                and name + ".linear.qweight" not in loaded_state_dict_keys_set
+            ):
+                continue
+
+            # insert MulLinear module
+            if name + ".linear.qweight" in loaded_state_dict_keys_set:
+                new_module = MulLinear(module)
+                set_module(model, name, new_module)
+                name += ".linear"
+
+            # replace `torch.nn.Linear` with `WeightOnlyLinear`
+            zp = True if name + ".qzeros" in loaded_state_dict_keys_set else False
+            g_idx = True if name + ".g_idx" in loaded_state_dict_keys_set else False
+
+            kwargs = {}
+            if _is_autoround:
+                from auto_round.export.export_to_itrex.model_wrapper import (
+                    WeightOnlyLinear as AutoRoundWeightOnlyLinear,
+                )
+
+                from .utility import convert_dtype_str2torch
+
+                WeightOnlyLinearClass = AutoRoundWeightOnlyLinear
+                kwargs["groupsize"] = module_quantization_config.get("group_size", 32)
+                kwargs["scale_dtype"] = convert_dtype_str2torch(module_quantization_config.get("scale_dtype", "fp16"))
+            else:
+                from .modules import WeightOnlyLinear as INCWeightOnlyLinear
+
+                WeightOnlyLinearClass = INCWeightOnlyLinear
+                kwargs["group_size"] = module_quantization_config.get("group_size", 32)
+                kwargs["g_idx"] = g_idx
+
+            new_module = WeightOnlyLinearClass(
+                module.in_features,
+                module.out_features,
+                dtype=module_quantization_config.get("dtype", "int"),
+                bits=module_quantization_config.get("bits", 4),
+                zp=zp,
+                bias=module.bias is not None,
+                use_optimum_format=True,
+                **kwargs,
+            )
+            set_module(model, name, new_module)
+
+    return model
+
+
+def _load_inc_woq_model(qmodel_weight_file_path, qconfig_file_path, origin_model):
+    qweights = torch.load(qmodel_weight_file_path)
+
+    quantization_config = {}
+    with open(qconfig_file_path, "r") as file:
+        quantization_config = json.load(file)
+
+    model = _build_woq_model(origin_model, quantization_config, qweights.keys())
+    model.load_state_dict(qweights, assign=True)
+    model.eval()
+    return model
+
+
+def _load_hf_woq_model(pretrained_model_name_or_path, *model_args, **kwargs):
```

Comment:
how about splitting this function into several parts? we will convert format from autogptq to hpu in this function and currently this function is too long to understand.

Reply:
updated in https://github.com/intel/neural-compressor/pull/1826/commits/939a261dd4554ea6e45dc2ae23586ab7eda4ed68